lib/syskit/roby_app/log_transfer_server/ftp_upload.rb

```diff
@@ -0,0 +1,111 @@
+# frozen_string_literal: true
+
+require "syskit/roby_app/log_transfer_server/log_upload_state"
+
+module Syskit
+    module RobyApp
+        module LogTransferServer
+            # Encapsulation of the log file upload process
+            class FTPUpload
+                def initialize( # rubocop:disable Metrics/ParameterLists
+                    host, port, certificate, user, password, file,
+                    max_upload_rate: Float::INFINITY,
+                    implicit_ftps: false
+                )
+
+                    @host = host
+                    @port = port
+                    @certificate = certificate
+                    @user = user
+                    @password = password
+                    @file = file
+
+                    @max_upload_rate = Float(max_upload_rate)
+                    @implicit_ftps = implicit_ftps
+                end
+
+                # Create a temporary file with the FTP server's public key, to pass
+                # to FTP.open
+                #
+                # @yieldparam [String] path the certificate path
+                def with_certificate
+                    Tempfile.create do |cert_io|
+                        cert_io.write @certificate
+                        cert_io.flush
+                        yield(cert_io.path)
+                    end
+                end
+
+                # Open the FTP connection
+                #
+                # @yieldparam [Net::FTP]
+                def open
+                    with_certificate do |cert_path|
+                        Net::FTP.open(
+                            @host,
+                            private_data_connection: false, port: @port,
+                            implicit_ftps: @implicit_ftps,
+                            ssl: { verify_mode: OpenSSL::SSL::VERIFY_PEER,
+                                   ca_file: cert_path }
+                        ) do |ftp|
+                            ftp.login(@user, @password)
+                            yield(ftp)
+                        end
+                    end
+                end
+
+                # Open the connection and transfer the file
+                #
+                # @return [LogUploadState::Result]
+                def open_and_transfer(root)
+                    open { |ftp| transfer(ftp, root) }
+                    LogUploadState::Result.new(@file, true, nil)
+                rescue StandardError => e
+                    LogUploadState::Result.new(@file, false, e.message)
+                end
+
+                def ensure_dataset_path_exists(ftp, root)
+                    dataset_path = File.dirname(@file.relative_path_from(root))
+
+                    dataset_path.split("/") do |folder|
+                        ftp.chdir(folder)
+                    rescue StandardError => e
+                        ftp.mkdir(folder)
+                        ftp.chdir(folder)
```

Comment:
Do not rescue this, its better to look for the folder instead, I guess using ftp.ls

Reply:
Actually, after taking a deeper look into this, using ls would make WriteOnlyFileSystem perform a read...and it would also increase the number of ftp requests per candidate to at least 2. I now think that rescuing the chdir is better.